### PR TITLE
Added function util_alloc_normal_path()

### PR DESCRIPTION
**Pull request by joakim-hove**
**Task**
Added a function which will normalize a path like: "../../path/subpath/./../../". Convenient for testing in libres.

**Pre un-WIP checklist**
- [ ] Statoil tests pass locally

### CODE REVIEW

**Comment by pgdr** [comment]
> Convenient for testing in libres

Perhaps it should be in _libres_  then?

File: lib/util/util.c [hunk 1]
```diff
@@ -895,6 +895,13 @@ char * util_alloc_realpath__(const char * input_path) {
         if (path_index == path_len)
           break;
 
+        /* We have backrefed us out of the current filesystem. */
+        if (prev_index < 0) {
+          free( abs_path );
+          free( real_path);
+          return NULL;
```
Comment (jokva) [inline lib/util/util.c]: Is this _really_ desired behaviour? That it returns `nullptr` when it hits "beyond root"?

Comment (pgdr) [reply lib/util/util.c]: I agree that I would expect `../../../../../../../..adinfinitum` to return root since `/.. = /`.

Comment (joakim-hove) [reply lib/util/util.c]: OK - I just checked this:
```
cd /tmp/path
cd ../../../../../../tmp/path
pwd 
/tmp/path
```
I must say I was a little surprised by that - but then yes I fully agree that the semantics should be changed; thank you.

File: lib/util/tests/ert_util_normal_path.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  test_work_area_type * work_area = test_work_area_alloc("Work");
+  util_make_path("level0/level1/level2");
+
+
+  test_path( "level0/level1/../", "level0");
+  test_path( "level0", "level0");
+  test_path( "level0/././level1/../", "level0");
+  test_path( "level0/level1/../level1/level2/../", "level0/level1");
+  test_path( "level0/level1/../level1/level2/..", "level0/level1");
+  test_path( "level0/level1/../level1/level2/../file.txt", "level0/level1/file.txt");
```
Comment (markusdregi) [inline lib/util/tests/ert_util_normal_path.c]: Please add a test that does not move linearly in the file system. Something like:
`test_path( "level0/level1/../level1a", "level0/level1a" );`
I expect it to pass, but I would like it to there.

File: lib/util/tests/ert_util_normal_path.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  free( normal_path );
+}
+
+
+void test_relative() {
+  test_work_area_type * work_area = test_work_area_alloc("Work");
+  util_make_path("level0/level1/level2");
+
+
+  test_path( "level0/level1/../", "level0");
```
Comment (markusdregi) [inline lib/util/tests/ert_util_normal_path.c]: Do we expect this to work at Windows as well? In that case I would appreciate a test with a Windows back slash path as well.

File: lib/util/tests/ert_util_normal_path.c (hunk 2 above)
Comment (joakim-hove) [reply lib/util/tests/ert_util_normal_path.c]: New test added and force-pushed

File: lib/util/tests/ert_util_normal_path.c (hunk 3 above)
Comment (joakim-hove) [reply lib/util/tests/ert_util_normal_path.c]: > Do we expect this to work at Windows as well?

Ehhh yes but no; we *expect* it to work on Windows, but we do not run the tests on Windows - so how naive is that?!

The way windows path support is implemented is that all operations working with paths use the symbol `UTIL_PATH_SEP_STRING` - and that resolves to `"/"` and `"\"` on Posix and Windows respectively. So - although the code is assumed to be portable to windows - and compile-tested for that - the tests are not written portably.

File: lib/util/util.c [hunk 4]
```diff
@@ -889,26 +890,33 @@ char * util_alloc_realpath__(const char * input_path) {
...
           break;
 
         if (strcmp(path_list[path_index] , BACKREF ) == 0) {
           mask[path_index] = false;
-          mask[prev_index] = false;
-          prev_index--;
           path_index++;
+          if (depth > 0) {
+            mask[prev_index] = false;
+            prev_index--;
```
Comment (markusdregi) [inline lib/util/util.c]: I believe there is a bug in this function (see #239). But it was here before this PR as well, so I just created an issue for it.

Comment (joakim-hove) [reply lib/util/util.c]: Yes - you are right; thank you for the thorough review.

File: lib/util/tests/ert_util_normal_path.c (hunk 2 above)
Comment (markusdregi) [reply lib/util/tests/ert_util_normal_path.c]: Btw, are these tests ran at Windows as well? I would expect them to fail due to the forward slash and `util_path_sep` using `UTIL_PATH_SEP_STRING` to split files.

Comment (joakim-hove) [reply lib/util/tests/ert_util_normal_path.c]: Ehh - no; as commented here: https://github.com/Statoil/libecl/pull/232#discussion_r149655688

**Comment by joakim-hove** [comment]
OK - implementation completely rewritten; should fix #239

File: lib/util/tests/ert_util_normal_path.c (hunk 3 above)
Comment (markusdregi) [reply lib/util/tests/ert_util_normal_path.c]: Yeah, that is an interesting approach. But beyond the scope of this PR if it is ever to be fixed.

**Comment by joakim-hove** [comment]
> I would be very happy if you would make a test out of the example I made

I have done that.

File: lib/util/util.c [hunk 5]
```diff
@@ -875,78 +875,67 @@ char * util_alloc_realpath__(const char * input_path) {
...
+      path_stack[i] = NULL;
 
     {
-      int path_index = 1;  // Path can not start with ..
-      int prev_index = 0;
-      while (true) {
-        if (path_index == path_len)
-          break;
+      int stack_size = 0;
+      int path_index = 0;
```
Comment (markusdregi) [inline lib/util/util.c]: This path_index needs to go for libecl to compile.

**Pull request merged**